File: languages/scala/src/provider/SummaryAnalysis.scala
package provider

import model.{TableAnalysis, UserInfo}

/**
  * Classe para analise de dados, obtem os valores máximo, minimo e media de valores
  * @see model.TableAnalysis
  */
class SummaryAnalysis extends TableAnalysis[Array[Double]]{

  /**
    * Realiza um resumo dos dados
    * @param userInfoList Lista de dados a ser analisada
    * @return Valores maximo, minimo e media
    * @see TableAnalysis
    */
  override def analysis(userInfoList: List[UserInfo]): Array[Double] = {

    var max = Double.MinValue
    var min = Double.MaxValue
    var sum: Double = 0.0

    userInfoList.foreach { userInfo =>
      sum += userInfo.credit
      if(userInfo.credit > max) max = userInfo.credit
      if(userInfo.credit < min) min = userInfo.credit
    }

    val mean: Double = sum/userInfoList.length

    return Array(max, min, mean)
  }
}

File: languages/kotlin/src/provider/SummaryAnalysis.kt
package provider

import model.UserInfo
import model.TableAnalysis
import model.exception.InvalidParameterException

/**
 * Classe para analise de dados, obtem os valores maximos, minimos e media dos creditos dos usuarios
 * @see model.TableAnalysis
 */
class SummaryAnalysis: TableAnalysis<Array<Double>> {

    /**
     * Realiza um resumo dos dados
     * @param userInfoList Lista de dados a ser analisada
     * @return Resumo dos valores dos creditos
     * @see TableAnalysis
     */
    override fun analysis(userInfoList: List<UserInfo>): Array<Double> {
        if (userInfoList.isEmpty())
            throw InvalidParameterException("'userInfoList' é vazio")

        var value = 0.0
        var min: Double = Double.MAX_VALUE
        var max: Double = Double.MIN_VALUE

        userInfoList.forEach {
            value += it.credit
            if(it.credit > max) max = it.credit
            if(it.credit < min) min = it.credit
        }

        val mean = value/userInfoList.size

        return doubleArrayOf(min, max, mean).toTypedArray()
    }
}